jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/ObtainBaseImageLayerStep.java

```diff
@@ -69,38 +107,50 @@
   private final BuildConfiguration buildConfiguration;
   private final ProgressEventDispatcher.Factory progressEventDispatcherFactory;
 
-  private final DescriptorDigest layerDigest;
+  private final Layer layer;
   private final @Nullable Authorization pullAuthorization;
+  private final BlobExistenceChecker blobExistenceChecker;
 
-  PullAndCacheBaseImageLayerStep(
+  ObtainBaseImageLayerStep(
       BuildConfiguration buildConfiguration,
       ProgressEventDispatcher.Factory progressEventDispatcherFactory,
-      DescriptorDigest layerDigest,
-      @Nullable Authorization pullAuthorization) {
+      Layer layer,
+      @Nullable Authorization pullAuthorization,
+      BlobExistenceChecker blobExistenceChecker) {
     this.buildConfiguration = buildConfiguration;
     this.progressEventDispatcherFactory = progressEventDispatcherFactory;
-    this.layerDigest = layerDigest;
+    this.layer = layer;
     this.pullAuthorization = pullAuthorization;
+    this.blobExistenceChecker = blobExistenceChecker;
   }
 
   @Override
-  public CachedLayerAndName call() throws IOException, CacheCorruptedException {
+  public PreparedLayer call() throws IOException, CacheCorruptedException, RegistryException {
+    DescriptorDigest layerDigest = layer.getBlobDescriptor().getDigest();
     try (ProgressEventDispatcher progressEventDispatcher =
             progressEventDispatcherFactory.create("checking base image layer " + layerDigest, 1);
         TimerEventDispatcher ignored =
             new TimerEventDispatcher(
                 buildConfiguration.getEventHandlers(), String.format(DESCRIPTION, layerDigest))) {
+
+      Optional<Boolean> layerExists = blobExistenceChecker.exists(layerDigest);
```

Comment:
What if we inverted this checker to be _blobRequiredChecker_?  Then we're removing any knowledge here about existence — which is frankly a bit odd since it may not exist locally, but the code here is for downloading locally.  And then I think it can just be a `Predicate` as the force-download will always return `true`.  `PreparedLayer` would have a `existsLocally` instead of `existsInTarget`, and `PushBlob*` would skip pushing blobs if they didn't exist locally (which makes sense).

Reply:
*sigh* I swear I did post a long comment, but it got lost and nowhere to find. I will update later.

Reply:
> `PushBlob*` would skip pushing blobs if they didn't exist locally (which makes sense).

The catch is this: say we have

1. B_LOCAL_REMOTE: base layer that exists locally and remotely
2. A_LOCAL_REMOTE: app layer that exists locally and remotely
3. B_LOCAL: base layer that exists locally but not remotely
4. A_LOCAL: app layer that exists locally but not remotely
5. B_REMOTE: base layer that exists remotely but not locally

then the table should be

| layer | eventually pushed | `BlobPusher` should call |
| --- | --- | --- |
| B_LOCAL_REMOTE | no | nothing |
| A_LOCAL_REMOTE | no | `checkBlob` |
| B_LOCAL | yes | `pushBlob` |
| A_LOCAL | yes | `checkBlob` and `pushBlob` |
| B_REMOTE | no | nothing |

Note B_* (base layers) and A_* (app layers) work differently (i.e., calling `checkBlob` or not). `BlobPusher` should be able to differentiate them do proper actions.

Reply:
@briandealwis does this make sense?

Reply:
Sorry, missed this.  Yes, that makes sense.  We need to test whether B is REMOTE to know if we need to download it. so as to be able to push it later.